### PR TITLE
brightness

### PR DESCRIPTION
**Pull request by StefaanVanDooren**
Made brightness global across all modes.
Added colorWipe Mode

### CODE REVIEW

File: Super_Simple_RGB_WiFi_Lamp/Framework.ino [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      fill_solid(ledString, wipePosition,CRGB( 0, 0, 0));
+      
+      if (wipePosition == NUM_LEDS) {
+        TurningOn = true;
+        wipePosition = -1;
+        
+      }
+     }  
+   }
+}
```
Comment (thebigpotatoe) [inline Super_Simple_RGB_WiFi_Lamp/Framework.ino:39]: Need to merge these methods into the LEDs.ino file to keep things in the same place.

File: Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino [hunk 2]
```diff
@@ -29,23 +29,25 @@
...
+int leftLeds[]    = {96,97,98,99,100,101};
+int rightLeds[]   = {30,31,32,33,34,35};
 
 // Eneter your wifi credentials here - If you would like to enter your wifi credentials now you can with these variables. This is a nice easy 
 // method to get your ESP8266 connected to your network quickly. If you don't you can always set it up later in the wifi portal.
 String SSID = "";
 String Password = "";
+
+bool autoOnWithModeChange = false;
 // ########################################################## End of Sketch Variables ##########################################################
```
Comment (thebigpotatoe) [inline Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino:51]: All the changes to the base variables needs to be reverted to remain the same as my tutorial

File: Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino [hunk 3]
```diff
@@ -149,6 +152,17 @@ uint8_t visualiserFadeUp                    = 32;
 uint8_t visualiserFadeDown                  = 32;
 uint8_t visualiserHueOffset                 = 170;
 
+//colorWipe
+int wipePosition = -1;
+bool TurningOn = true;
+int colorWipeRed   = 255;
+int colorWipeGreen = 0;
+int colorWipeBlue  = 255;
+int wipeSpeed = 20;
```
Comment (thebigpotatoe) [inline Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino:161]: make sure each variable for the individual modes has the mode name at the start of the variable. ie; colourWipeTurningOn

File: Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino [hunk 4]
```diff
@@ -149,6 +152,17 @@ uint8_t visualiserFadeUp                    = 32;
...
+
+int SparklePixel;
+bool SparkleActive = true;
```
Comment (thebigpotatoe) [inline Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino:164]: Keep it camel case

**Comment by thebigpotatoe** [comment]
Had a read through your code, and haven't tested anything, but I think you should split the two features in this branch into separate ones. This will make it easier to track in the long run and easier to understand which commits relate to each feature. Essentially keeping 1 feature to 1 pull. Have a read of the comments. I made a few which need to be tidied up before merging in. But first I would have a read of the below.

I really like the looks of the colour wipe mode, always great to see some new modes contributed. But the global brightness idea has both some positive and negatives for implementation with the way I have written my previous modes. 

I have read over your code and mine quite a few times with regards this, and from what I can see it contradicts a couple of things. The first is the way the colour's are set for each mode that uses colour selection (ie colour, clock, and bell curve). In essence, you set each RGB channel which technically handles the brightness component at the same time as colour selection. Since you can set each of the channels to a value of 0 to 255 a dim red light could be CRGB(64, 0, 0) set from the RGB picker of the website for example. This allows the user to use a colour picker like the one I have chosen to also adjust the brightness as well as colour at the same time. 

The issue with a global brightness over the top of RGB values, and the way FastLED would handle it, means that an already dim light will be dimmed further. This could lead to confusing or undesirable results when trying to use the light as there are essentially two brightness values you will have to adjust to get the brightness you want.

The second issue would be if someone else were to write a mode which they did not want global dimming to occur, such as my night rider mode. I wrote this to be simple and just do one thing at a set brightness. Decreasing its brightness would affect the way it looks and therefore defeat what I was trying to achieve. Again this might not be the end of the world but it does in the long run restrict the implementation of new modes. Instead I encourage new modes to use their own brightness variables. These will then be saved in flash and reloaded each time a mode is changed allowing for a more dynamic setting.

The global brightness would also affect the way the visualiser brightness is set so I would not want it to be applied during that mode.

I understand why you would want to use a global brightness in modes such as rainbow and new modes such as the ones you have written. It also makes sense from a control point of view by being able to send the brightness as a single message rather than a mode with the correct values. But hopefully from the explanation above you can understand why I have not done so in the first place. 

I absolutely encourage you to use my code in any way that you want and to modify and play with it till you are happy, but I think for the meantime I will have to pass on the global brightness setting in the main branch. If you can figure out a decent way of being able to implement it without affecting colour accuracy or mode behavior I am all ears. If in the mean time you need help merging in the new modes into a pull request I can give you a hand.

File: Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino (hunk 2 above)
Comment (StefaanVanDooren) [reply Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino:51]: Can't those variables be put in a seperate file, for instance "UserSettings"? Would make it easier to maintain I think.
Also I believe the autoOnWithModeChange has to be placed there.... (putted back to true)

File: Super_Simple_RGB_WiFi_Lamp/Framework.ino (hunk 1 above)
Comment (StefaanVanDooren) [reply Super_Simple_RGB_WiFi_Lamp/Framework.ino:39]: Will do that.

File: Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino (hunk 3 above)
Comment (StefaanVanDooren) [reply Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino:161]: Ok, fair enough.

File: Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino (hunk 4 above)
Comment (StefaanVanDooren) [reply Super_Simple_RGB_WiFi_Lamp/Super_Simple_RGB_WiFi_Lamp.ino:164]: I will have to look that one up..... camel case...

**Comment by StefaanVanDooren** [comment]
I understand your concern (I think) and I'll see if I can come up with a solution. Being said, I do like to be able to set the Knight Rider mode a bit more dimmed (doesn't affect the coolness in anyway I think).
I can't comment on the visualiser, since I don't have anything connected to the analogue input yet...

**Comment by thebigpotatoe** [comment]
Understandable. Had a good think in the car on the way home. It might all be achievable with a method like maximizeBrightnes from fastLED.

**Comment by StefaanVanDooren** [comment]
You may close this pull request. At the moment I'm working on a whole new program, combining your work with a few others. I'll let you know when I have a working example.

**Comment by thebigpotatoe** [comment]
No worries. Thanks for your efforts. Let us know when you are done to see if there is a chance for merging.

**Pull request closed without merging**